### ISSUE TITLE
Model.eval() get more and more gpu memory

### ISSUE THREAD

**Issue by bilipa** [pytorch/pytorch#4932]
PyTorch GitHub Issues Guidelines
--------------------------------

We like to limit our issues to bug reports and feature requests. If you have a question or would like help and support, please visit our forums: https://discuss.pytorch.org/

If you are submitting a feature request, please preface the title with [feature request].

When submitting a bug report, please include the following information (where relevant):
- OS:
- PyTorch version:'0.4.0a0+33bb849'
- How you installed PyTorch (conda, pip, source): source
- Python version: 3.5
- CUDA/cuDNN version: 8.0/7.0
- GPU models and configuration:
- GCC version (if compiling from source):

In addition, including the following information will also be very helpful for us to diagnose the problem:
- A script to reproduce the bug. Please try to provide as minimal of a test case as possible.
- Error messages and/or stack traces of the bug
- Context around what you are trying to do

I use this example https://github.com/pytorch/examples/blob/master/imagenet/main.py
but I found in validation phase, the gpu get more and more memory.

**Comment by SsnL** [comment]
I believe that the issue is due to `volatile=True` having no effect on master and being replaced by `with torch.no_grad():`. The script should be updated before/at next release. See https://github.com/pytorch/examples/issues/280.

**Comment by jekbradbury** [comment]
An attempt to set a variable's `volatile` flag should maybe be a hard error on 0.4, since it always signifies code that needs to be updated to use `torch.no_grad`.

**Comment by bilipa** [comment]
THX.
I search this **torch.no_grad** in pytorch forums, and
I follow this: https://discuss.pytorch.org/t/torch-no-grad/12296, then my code is work.

**Comment by ezyang** [comment]
@jekbradbury I agree with you!

**Comment by Hayao41** [comment]
@SsnL Hi, thx for solution, but i don't understand why evaluation stage will use more memory than train stage when i use the same data as training also validating.

**Comment by SsnL** [comment]
@Hayao41 Because the computation graph is still created in this case, things like accumulating losses (in tensors) will keep them alive and thus use memory.

**Comment by Hayao41** [comment]
@SsnL Thx very much! According to your answer, dose the computation graph will not be freed in evaluation loop? My evaluation loop like following:
```
def eval():
    model.eval()
    output = model(input)
    loss = loss_func(output, target)
    # metric computing
```
So in this loop, loss will not be accumulated but the cuda memory usage is bigger than training loop with the same data and batch size.

**Comment by SsnL** [comment]
@Hayao41 Many things can happen, e.g., the allocate caches more memory, some training graph isn't freed, the model itself just uses more memory in eval, etc. You should just wrap the function in no_grad().

**Comment by Hayao41** [comment]
@SsnL Thx! I got it
